### PR TITLE
HPCC-17759 Windows build failure http.cpp

### PR DESCRIPTION
**Pull request by mckellyln**
Signed-off-by: Mark Kelly <mark.kelly@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

various soapplus tests successful

### CODE REVIEW

**Comment by mckellyln** [comment]
@mayx needed to change strndup() to StringBuffer for Windows build.  Unrelated, but took opportunity to add finer control for printing xml validation with stress tests.  Please review.
@xwang2713 fyi

File: esp/tools/soapplus/http.cpp [hunk 1]
```diff
@@ -1411,17 +1411,16 @@ int HttpClient::validate(StringBuffer& xml)
         try
         {
             Owned<IPropertyTree> testTree = createPTreeFromXMLString(len, bptr, ipt_caseInsensitive);
-            fprintf(m_logfile, "Successfully parsed XML\n");
+            if ( (!m_doStress) || (m_doStress && http_tracelevel >= 5) )
+                fprintf(m_logfile, "Successfully parsed XML\n");
```
Comment (mckellyln) [inline esp/tools/soapplus/http.cpp]: unrelated to Windows build error, but took opportunity to add finer control of printing validation with --stress option for better output.

File: esp/tools/soapplus/http.cpp [hunk 2]
```diff
@@ -1411,17 +1411,16 @@ int HttpClient::validate(StringBuffer& xml)
...
         }
         catch(IException *e)
         {
             StringBuffer emsg;
             fprintf(m_logfile, "Error parsng XML %s\n", e->errorMessage(emsg).str());
```
Comment (richardkchapman) [inline esp/tools/soapplus/http.cpp]: Typo: parsng

File: esp/tools/soapplus/http.cpp [hunk 3]
```diff
@@ -1411,17 +1411,16 @@ int HttpClient::validate(StringBuffer& xml)
...
             // is it ok to say bptr[len] = '\0' ?
-            char *newxml = strndup(bptr, len);
-            newxml[len] = '\0';
-            fprintf(m_logfile, "result xml:\n%s\n\n", newxml);
```
Comment (ghalliday) [inline esp/tools/soapplus/http.cpp]: Could also use precision option for a string:

printf("%.*s", len, bptr)

Comment (mckellyln) [inline esp/tools/soapplus/http.cpp]: @ghalliday yes, good suggestion.  I'll change.

**Comment by mckellyln** [comment]
@ghalliday, @mayx updated, thanks.

**Pull request merged**